### PR TITLE
Clean up argument checking in communication calls

### PR DESCRIPTION
**Pull request by bwbarrett**
Lots of argument checks are expensive in the critical path.  Refactor the argument checking so that we always validate inputs from NCCL, but use asserts for validating our internal data structure linking.

Since I'm moving the code around a bit, also change the error messages when group receives with an invalid 'n' are found.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/nccl_ofi_rdma.c [hunk 1]
```diff
@@ -3109,60 +3062,31 @@ static int recv(nccl_net_ofi_recv_comm_t *recv_comm, int n, void **buffers,
...
 
-	/* Currently, plugin doesn't support grouped receives */
-	if (n > NCCL_OFI_MAX_RECVS) {
-		NCCL_OFI_WARN("Group receives are not supported!");
-		ret = -ENOTSUP;
-		goto error;
-	}
+	int dev_id = r_comm->base.base.dev_id;
 
-	if (OFI_UNLIKELY(mr_handles == NULL)) {
```
Comment (rauteric) [inline src/nccl_ofi_rdma.c:3157]: It seems this check was lost. Should we put it in api.c?

File: src/nccl_ofi_rdma.c [hunk 2]
```diff
@@ -3109,60 +3062,31 @@ static int recv(nccl_net_ofi_recv_comm_t *recv_comm, int n, void **buffers,
...
-		NCCL_OFI_WARN("Memory handles array is NULL");
-		ret = -EINVAL;
-		goto error;
-	}
+	nccl_net_ofi_rdma_ep_t * ep = (nccl_net_ofi_rdma_ep_t *)r_comm->base.base.ep;
+	assert(ep != NULL);
 
-	if (mr_handles[0] == NULL) {
```
Comment (rauteric) [inline src/nccl_ofi_rdma.c:3163]: And this one?

File: src/nccl_ofi_rdma.c (hunk 1 above)
Comment (bwbarrett) [reply src/nccl_ofi_rdma.c:3157]: fixed

File: src/nccl_ofi_rdma.c (hunk 2 above)
Comment (bwbarrett) [reply src/nccl_ofi_rdma.c:3163]: fixed

**Review by rauteric — APPROVED** [review]

**Review by rauteric — APPROVED** [review]

**Comment by bwbarrett** [comment]
One more try.  I shouldn't have let @rauteric talk me into checking the value of mr handles, because it turns out that we do return NULL as the handle when Libfabric doesn't require local registration.

**Review by rauteric — APPROVED** [review]

**Comment by bwbarrett** [comment]
P3dn test infrastructure is broken, so ignoring and committing this PR.

**Pull request merged**